File: Modules/ServerTypes/EmployeeAttendance/ShiftAttendanceRecordForm.ts
﻿import { LookupEditor, DateEditor, StringEditor, TextAreaEditor, PrefixedContext } from "@serenity-is/corelib";
import { initFormType } from "@serenity-is/corelib/q";

export interface ShiftAttendanceRecordForm {
    EmployeeRowId: LookupEditor;
    ShiftId: LookupEditor;
    TimeIn: DateEditor;
    TimeInHour: StringEditor;
    TimeOut: DateEditor;
    TimeOutHour: StringEditor;
    ShiftStartTime: DateEditor;
    ShiftStartTimeHour: StringEditor;
    ShiftEndTime: DateEditor;
    ShiftEndTimeHour: StringEditor;
    EmpRemark: TextAreaEditor;
    SupRemark: TextAreaEditor;
    LvPhRemark: TextAreaEditor;
}

export class ShiftAttendanceRecordForm extends PrefixedContext {
    static formKey = 'EmployeeAttendance.ShiftAttendanceRecord';
    private static init: boolean;

    constructor(prefix: string) {
        super(prefix);

        if (!ShiftAttendanceRecordForm.init)  {
            ShiftAttendanceRecordForm.init = true;

            var w0 = LookupEditor;
            var w1 = DateEditor;
            var w2 = StringEditor;
            var w3 = TextAreaEditor;

            initFormType(ShiftAttendanceRecordForm, [
                'EmployeeRowId', w0,
                'ShiftId', w0,
                'TimeIn', w1,
                'TimeInHour', w2,
                'TimeOut', w1,
                'TimeOutHour', w2,
                'ShiftStartTime', w1,
                'ShiftStartTimeHour', w2,
                'ShiftEndTime', w1,
                'ShiftEndTimeHour', w2,
                'EmpRemark', w3,
                'SupRemark', w3,
                'LvPhRemark', w3
            ]);
        }
    }
}

File: Modules/HumanResource/EmployeeAttendance/ShiftAttendanceRecord/ShiftAttendanceRecordForm.cs
using Serenity.ComponentModel;
using System;
using System.ComponentModel;

namespace HRMSoftware.EmployeeAttendance.Forms;

[FormScript("EmployeeAttendance.ShiftAttendanceRecord")]
[BasedOnRow(typeof(ShiftAttendanceRecordRow), CheckNames = true)]
public class ShiftAttendanceRecordForm
{
    public int EmployeeRowId { get; set; }
    public int ShiftId { get; set; }
    [Category("Time In")]

    [HalfWidth]
    public DateTime TimeIn { get; set; }
    [HalfWidth]
    public string TimeInHour { get; set; }
    [Category("Time Out")]

    [HalfWidth]
    public DateTime TimeOut { get; set; }
    [HalfWidth]
    public string TimeOutHour { get; set; }
    [Category("Shift Start")]
    [HalfWidth]
    public DateTime ShiftStartTime { get; set; }
    [HalfWidth]
    public string ShiftStartTimeHour { get; set; }
    [Category("Shift End")]

    [HalfWidth]
    public DateTime ShiftEndTime { get; set; }
    [HalfWidth]
    public string ShiftEndTimeHour { get; set; }
    [TextAreaEditor(Rows = 3)]
    public string EmpRemark { get; set; }
    [TextAreaEditor(Rows = 3)]
    public string SupRemark { get; set; }
    [TextAreaEditor(Rows = 3)]
    public string LvPhRemark { get; set; }

    //public int ShiftId { get; set; }
    /*
    public int LateIn { get; set; }
    public int Ot { get; set; }
    public int EarlyLeave { get; set; }
    public int TimeInRowId { get; set; }
    public int TimeOutRowId { get; set; }
    */
    //public DateTime ShiftDate { get; set; }
}